Add GCP destination

mitchcail opened the pull request:
This PR sets up a destination plugin for GCP and will allow us to upload certificates to our GCP datacenters. 
It allows 2 ways to authenticate API requests. One is via vault (This is the option we will be using at DD) and the other is via service account tokens. 
This PR is currently running in our sandbox environment.

henrymwang commented on lemur/plugins/lemur_gcp/readme.md:
```diff
@@ -0,0 +1,18 @@
+When setting up the new destination in the lemur console set the "using vault" value to True. 
+Then enter the path to your secret credentials. Lemur will use this token to authenticate API requests.
+
+#### Authentication by Service Account Credentials
```
Can you add docs for how to auth locally with the gcloud cli?

mitchcail replied:
Ya I can do that

mitchcail replied:
updated with instructions to run locally

mitchcail commented on docker/src/lemur.conf.py:
```diff
@@ -32,7 +32,7 @@ def get_random_secret(length):
 # This must match the key for whichever DB the container is using - this could be a dump of dev or test, or a unique key
 LEMUR_ENCRYPTION_KEYS = repr(os.environ.get('LEMUR_ENCRYPTION_KEYS',
-                                            base64.b64encode(get_random_secret(32).encode('utf8')).decode('utf8')))
+                                            base64.b64encode(b'SBRHNTIW@%#$)_#+ulahitit47774808').decode('utf8')))
```
This fixes a bug we have locally where the `LEMUR_ENCRYPTION_KEYS` gets rotated and suddenly queries to the DB no longer work because of a key mismatch. This lemur.conf file is only used for local testing.

mitchcail commented on docker/src/lemur.conf.py:
```diff
@@ -232,3 +232,5 @@ def get_random_secret(length):
 # Authority plugins which support revocation
 SUPPORTED_REVOCATION_AUTHORITY_PLUGINS = ['acme-issuer']
+
+# PATH_TO_GOOGLE_APPLICATION_CREDENTIALS = '/tmp/authentication.json'
```
I think its a nice to have. In order to test locally you would uncomment this (or change the value based on where the credential file is stored in your container). It also follows a similar pattern of some other variables we have commented out in this file, then if you want to test locally you uncomment.

bobmshannon replied:
Suggest making this an option which can be specified directly in the destination plugin option instead, for example using an option called `tokenFilePath`. (Consider the use-case where you have different credentials for different GCP projects and one destination plugin configured for each project.)

bobmshannon commented on lemur/plugins/lemur_gcp/plugin.py:
```diff
@@ -0,0 +1,99 @@
+            "helpMessage": "GCP Service Account Name",
+        },
+        {
+            "name": "Use Vault",
```
Suggest calling this `authenticationMethod` with valid enum values of `vault` or `tokenFile` to be consistent with patterns established by other plugins. ([example](https://github.com/DataDog/lemur/blob/master/lemur/plugins/lemur_vault_dest/plugin.py#L55))

bobmshannon commented on lemur/plugins/lemur_gcp/plugin.py:
```diff
@@ -0,0 +1,99 @@
+
+    def _get_gcp_credentials(self, options):
+
+        if self.get_option('Use Vault', options) is True:
```
```suggestion
        if self.get_option('Use Vault', options):
```

bobmshannon commented on lemur/plugins/lemur_gcp/plugin.py:
```diff
@@ -0,0 +1,99 @@
+    author = "Mitch Cail"
+    author_url = "https://github.com/Datadog/lemur"
+
+    options = [
```
Side note (no action required): These option names get converted to title case in the frontend which means an option named `authenticationMethod` will be displayed as `Authentication Method` in the UI.

mitchcail replied:
Nice! I'll switch out for camel case 🐫

bobmshannon commented on lemur/plugins/lemur_gcp/plugin.py:
```diff
@@ -0,0 +1,107 @@
+            }
+            credentials = self._get_gcp_credentials(options)
+            return self._insert_gcp_certificate(
+                self.get_option("AccountId", options),
```
This `AccountId` option doesn't seem to exist (the lookup is case-sensitive) so I don't think this will work properly. (We can and should add a test case for this.)

bobmshannon commented on lemur/plugins/lemur_gcp/plugin.py:
```diff
@@ -0,0 +1,107 @@
+
+    options = [
+        {
+            "name": "accountId",
```
Nit: Call this `projectID` to match GCP nomenclature?

henrymwang replied:
Noting here the frontend parsing logic makes it render as "Project I D".

bobmshannon commented on lemur/plugins/lemur_gcp/tests/test_plugin.py:
```diff
@@ -0,0 +1,89 @@
+}
+
+options = [{
+    'name': 'accountName',
```
This isn't a valid option for this plugin and can be removed or more preferably replaced with valid options which are tested appropriately, see next comment below.

mitchcail replied:
This makes sense I can make that change!

bobmshannon commented on lemur/plugins/lemur_gcp/tests/test_plugin.py:
```diff
@@ -0,0 +1,89 @@
+@mock.patch("lemur.plugins.lemur_gcp.plugin.GCPDestinationPlugin.get_option", return_value="test-account")
+@mock.patch("lemur.plugins.lemur_gcp.plugin.GCPDestinationPlugin._get_gcp_credentials", return_value=token)
+@mock.patch("lemur.plugins.lemur_gcp.plugin.GCPDestinationPlugin._insert_gcp_certificate", return_value=SUCCESS_INSERT_RESPONSE)
+def test_upload(mock_sslCertificates, mock_credentials, mock_gcp_acount_id):
```
I think we can use [assert_called_with](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_with) to assert that these mocked methods are being called with the correct arguments based on the provided options to the destination plugin. This approach should catch the case-sensitivity lookup issue with the `AccountId` option, for example.

mitchcail replied:
I  will look into this.

mitchcail replied:
I added the `assert_called_with` as part of this [PR](https://github.com/DataDog/lemur/pull/15/commits/767a6bf1a1a0ce8cf3de46a7e97f6cbbe76e4d34)

bobmshannon commented on lemur/plugins/lemur_gcp/tests/test_plugin.py:
```diff
@@ -0,0 +1,89 @@
+
+
+@mock.patch("lemur.plugins.lemur_gcp.plugin.GCPDestinationPlugin._get_gcp_credentials_from_vault", return_value="ya29.c.b0AXv0zTN36HtXN2cJolg9tAj0vGAOT29FF-WNxQzvPu")
+def test_get_gcp_credentials__from_vault(mock__get_gcp_credentials_from_vault):
```
Two minor nits:
- This method has two consecutive underscores in it (?)
- Double and single quotes seem to be randomly used within this file for strings

mitchcail replied:
Its a naming convention I picked up from previous teams i've been on to give a descriptor to the test. 
It follows a convention of `test_{method name}__{descriptor for test}`. I use it in pytest where you often have many separate tests for the same method so you can quickly get a descriptor for the given test. 
eg:
```
test_get_gcp_credentials__from_vault()
test_get_gcp_credentials__with_service_account_creds()
test_get_gcp_credentials__exception_thrown_without_valid_creds()
```
This isn't something defined in the pep8 style guide and is just a habit i've continued with. I'm happy to switch it so that matches our coding style. 

To you second bullet point, you are right, I accidentally have some double underscores in string names, eg: `mock__get_gcp_credentials_from_vault`, this was not intentional.

mitchcail commented:
I think I've addressed all the comments on the PR with the exception of writing the documentation for local testing. I will write this up next week when Im back.

bobmshannon commented on lemur/plugins/lemur_gcp/plugin.py:
```diff
@@ -0,0 +1,106 @@
+            roleset="",
+            mount_point=f"{self.get_option('vaultMountPoint', options)}"
+        )["data"]["token"].rstrip(".")
+        credentials, _ = google.auth.default()  # Fetch the default credentials from Emissary Native IAM
```
Let's remove any references to internal tools in here -- I was thinking something like `Fetch credentials from Vault GCP secrets engine` for example.

mitchcail replied:
Agreed. Fix this in the latest commit.

bobmshannon reviewed (commented):
I think we can probably start with this especially since it's been tested on sandbox and has test coverage. However, I believe we'll still need a solution in order to ensure that the name of the certificates uploaded match GCP API's validation constraints.

bobmshannon commented on lemur/plugins/lemur_gcp/plugin.py:
```diff
@@ -0,0 +1,105 @@
+            roleset="",
+            mount_point=f"{self.get_option('vaultMountPoint', options)}"
+        )["data"]["token"].rstrip(".")
+        credentials, _ = google.auth.default()  # Fetch credentials from Vault GCP secrets engine
```
More accurately, this line is fetching the default credentials for the environment as-per https://google-auth.readthedocs.io/en/master/reference/google.auth.html. The call to `generate_oauth2_access_token` in the lines above fetch the OAuth2 token from Vault.

Ultimately how much we want to tweak these comments is up to you but as written it could be misleading for someone not familiar with GCP auth attempting to make edits to this plugin in the future.

mitchcail replied:
👍 updated this comment to closer reflect what is written in the google auth documentation.

bobmshannon commented on lemur/plugins/lemur_gcp/tests/test_plugin.py:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    # assert our mocks are being called with the params we expect
+    mock_sslCertificates.assert_called_with('lemur-test', ssl_certificate_body, token)
```
```suggestion
    mock_ssl_certificates.assert_called_with('lemur-test', ssl_certificate_body, token)
```

bobmshannon commented on lemur/plugins/lemur_gcp/readme.md:
```diff
@@ -0,0 +1,18 @@
+
+#### Authentication by Service Account Credentials
+In the GCP console generate a new set of Service Account Credentials. Store those on your server and then
+set the `PATH_TO_GOOGLE_APPLICATION_CREDENTIALS` env variable in the lemur.conf file to point to where that key is stored.
```
This is no longer accurate since the token file path is now configured directly in the destination plugin.

mitchcail commented:
> I believe we'll still need a solution in order to ensure that the name of the certificates uploaded match GCP API's validation constraints.

Thats a good call out. I actually forgot that was needed. I don't think it should be too much work and we can have it as part of this PR.

mitchcail commented on lemur/plugins/lemur_gcp/plugin.py:
```diff
@@ -0,0 +1,105 @@
+            mount_point=f"{self.get_option('vaultMountPoint', options)}"
+        )["data"]["token"].rstrip(".")
+        credentials, _ = google.auth.default()  # Fetch credentials from Vault GCP secrets engine
+        credentials.token = service_token  # replace the token from Native IAM with the Dataproc token fetched from Vault
```
updated the comment: https://github.com/DataDog/lemur/pull/15/commits/83474846e63865bb24dc0a9b54f387fe15848066

mitchcail commented:
I updated the plugin to include a method for certificate naming, since we need a way to name the certificate in a way that GCP expects.

bobmshannon commented on lemur/plugins/lemur_gcp/plugin.py:
```diff
@@ -0,0 +1,114 @@
+
+    def _gcp_name(self, body):
+        """
+        We need to change the name of the certificate that we are uploading to comply with GCP naming standards.
```
There's apparently two more GCP naming constraints which need to be handled here:

- certificate names must be all lower case (the `authority` in this case can contain upercase characters)
- no longer than 63 characters (names longer than 63 characters can be generated)

mitchcail replied:
added the new naming constraints as part of this commit: https://github.com/DataDog/lemur/pull/15/commits/e72d08d8d880ee926fa59d6d6c38c9b2f04c815f
I moved the logic for modifying the cert name to its own method for easier testing.

mitchcail replied:
follow up. Some uploads were still failing because trailing characters in the name must be alpha numeric. Fixed this issue with https://github.com/DataDog/lemur/pull/15/commits/f02f4743d318607bafa452cfeb36d01125889904
